launcher

bartoszkobylinski opened the pull request.

bartoszkobylinski committed:
Major features included:

1. Basic colors (WHITE, RED, BLUE) definitions for screen visuals.
2. `BaseView` class as a foundation to build other game views.
3. `MenuView` for rendering and handling the main game menu. This includes a button with label "Game Selection" that, upon mouse click, is intended to transition to a game selection view.
4. Game loop for processing events and updating the screen.

Note: The `GameSelectionView` is only partially implemented and serves as a placeholder for future enhancements.

bartoszkobylinski committed:
Title: Refactor menu.py
Message:
Transitioned from class-based views to function-based views in menu.py for better clarity and modularity.
Introduced utility functions to reduce code redundancy.
Enhanced game selection view to offer difficulty options.

busterbeam committed:
Update menu.py

edited in GitHub. the `__main__` guard ensures when imported it doesn't run `menu`, and instead can be called from a specific point in script. and `menu` should also return the "action"

busterbeam commented:
You'll need to fix some things before closing the branch. as the linting checks aren't being accepted.  `pre-commit` should of caught this. but also getting pyright again. you can ignore pyright for now.

flake8 is totally right about docstrings.  but besides that this code looks superb I'll test run it in an hour.

bartoszkobylinski committed:
Title: Implement Jigsaw game enhancements
Message:
Add Hamming distance calculation for puzzle positions.
Introduce a completion bar showing game progress.
Implement a tile position simulator.

busterbeam committed:
added some direction to `__main__.py` and `client/__init__.py` to allow actions from `menu` go to `GameClient`.  menu's colour palatte is a bit on the hard side and we do need to define a proper `action` dictionary.  Hopefully this will help

busterbeam commented on jigsaw-game/client/menu.py:
```diff
@@ -0,0 +1,110 @@
+def menu():
+    current_view = "menu"
+    current_difficulty = None
+    view_functions = {
+        "menu": menu_view,
+        "game_selection": game_selection_view,
+    }
+
+    while True:
+        if current_view == "play_game":
+            action = play_game(current_difficulty)
+        else:
+            action = view_functions[current_view]()
```
interesting usage of accessing functions.  maybe instead try `globals()[current_view]()`.. but maybe add a few lines, to check we are accessing a function such as `function = globals()[current_view] if callable(globals()[current_view]) else not_implemented_function`.

LukeHankey replied:
I would argue strongly against using `globals()`. It's a bad practice to get into a habit of. Simple if checks on all of them is fine, or to clean it up in one line, just have `action = view_functions[current_view]()` and add the `play_game` action to `view_functions`.

busterbeam replied:
:smirk: don't like the dangerous side? as long as we are checking that they are callable. also a simple startswith filter for dunders would probably fix the more extreme cases.  Either that, or really it should be using `function.__name__` in the key place so you don't have to update code everywhere when you change names.

busterbeam commented on jigsaw-game/client/menu.py:
```diff
@@ -0,0 +1,110 @@
+def game_selection_view():
+    screen.fill(WHITE)
+
+    easy_button = pygame.Rect(100, 100, 200, 50)
+    moderate_button = pygame.Rect(100, 200, 200, 50)
+    hard_button = pygame.Rect(100, 300, 200, 50)
```
for **Quick Game** Easy, Intermediate and Hard are good options, but for **Custom Game** we should include either text entry or better still a slider for how many "tiles"... but make it follow squared numbers 4, 9, 16, 25, 36, 49, 64, 81, etc.

busterbeam commented on jigsaw-game/client/menu.py:
```diff
@@ -0,0 +1,110 @@
+pygame.init()
+
+# Colors
+WHITE = (255, 255, 255)
+RED = (255, 0, 0)
+BLUE = (0, 0, 255)
+
+# Screen dimensions
+WIDTH, HEIGHT = 800, 600
+screen = pygame.display.set_mode((WIDTH, HEIGHT))
+pygame.display.set_caption('Game Menu')
+
+font = pygame.font.SysFont(None, 36)  # Default font, size 36
```
this should be in `menu` function and dimensions could be put into a `constants.py` file as `GameClient` uses it aswell, but we should resizeable, and this should probably a "configurable" option in "Settings".. do have fun with fonts

busterbeam reviewed (changes requested):
good start, hopefully this will help

bartoszkobylinski committed:
Enhancements and bugfixes in Jigsaw Game

    - Corrected a typo in the TODO list by fixing the spelling of 'compatibility'
    - Added new modules `constants.py`, `credits_view.py`, and `settings_view.py` to handle game constants, credits view, and settings view, respectively
    - Updated `menu.py` by adding functionality for easy, medium, hard game selection, and settings and credits view
    - Removed hardcoded colors and screen dimensions from `menu.py` and moved them to `constants.py`
    - Created GUI components for settings view but commented them out for future implementation

bartoszkobylinski committed:
Enhancements and bugfixes in Jigsaw Game

    - Corrected a typo in the TODO list by fixing the spelling of 'compatibility'
    - Added new modules `constants.py`, `credits_view.py`, and `settings_view.py` to handle game constants, credits view, and settings view, respectively
    - Updated `menu.py` by adding functionality for easy, medium, hard game selection, and settings and credits view
    - Removed hardcoded colors and screen dimensions from `menu.py` and moved them to `constants.py`
    - Created GUI components for settings view but commented them out for future implementation

bartoszkobylinski committed:
Merge remote-tracking branch 'origin/launcher' into launcher

bartoszkobylinski committed:
Enhancements and bugfixes in Jigsaw Game

    - Added new modules `constants.py`, `credits_view.py`, and `settings_view.py` to handle game constants, credits view, and settings view, respectively
    - Updated `menu.py` by adding functionality for easy, medium, hard game selection, and settings and credits view
    - Removed hardcoded colors and screen dimensions from `menu.py` and moved them to `constants.py`
    - Created GUI components for settings view but commented them out for future implementation

LukeHankey commented on jigsaw-game/client/buttons.py:
```diff
@@ -0,0 +1,47 @@
+import pygame
+import pygame_gui
```
Add to `requirements.txt`

LukeHankey commented on jigsaw-game/client/buttons.py:
```diff
@@ -0,0 +1,47 @@
+import pygame
+import pygame_gui
+from constants import HEIGHT, WIDTH
```
```suggestion
from .constants import HEIGHT, WIDTH
```

LukeHankey commented on jigsaw-game/client/menu.py:
```diff
@@ -167,7 +220,6 @@ def menu():
-if __name__ == "__main__":
-    print(menu())
+
+menu()
```
```suggestion
```

LukeHankey commented on jigsaw-game/client/settings_view.py:
```diff
@@ -0,0 +1,88 @@
+from constants import screen, WHITE, RED, BLUE, font, WIDTH, HEIGHT, manager
+
+#  Create a GUI components
+'''
```
What is this commented for?

LukeHankey commented on jigsaw-game/client/settings_view.py:
```diff
@@ -0,0 +1,88 @@
+import pygame_gui.elements
+from pygame import QUIT
+
+from constants import screen, WHITE, RED, BLUE, font, WIDTH, HEIGHT, manager
```
```suggestion
from .constants import screen, WHITE, BLUE, font, WIDTH, manager
```

busterbeam committed:
Update jigsaw-game/client/menu.py

Yes I changed this before why do you keep changing it back bart?!

Co-authored-by: Luke <l.hankey@oxcan.org>

PiLogic commented:
Hello! Is there any chance my username could be added to the Credits Page? Thank you, that would be greatly appreciated.

bartoszkobylinski committed:
Title:
Refactor button names and add custom game view
Description:
- Changed button names and texts in buttons.py to better reflect their functionality.
- Added a custom game view in custom_game.py and integrated it in the menu.py.
- Refactored settings_view.py to add an open file dialog and removed unused GUI components.

bartoszkobylinski committed:
Merge remote-tracking branch 'origin/launcher' into launcher
